File: backend/internal/intelligence/compute.go
package intelligence

import (
	"fmt"
	"math"
	"sort"
	"time"

	"github.com/google/uuid"
)

// ─── Forecast ─────────────────────────────────────────────────────────────────

// ForecastPoint is one projected month in the cash-flow forecast.
type ForecastPoint struct {
	Month            string  // "2026-06"
	ProjectedInflow  float64 `json:"projected_inflow"`
	ProjectedOutflow float64 `json:"projected_outflow"`
	ProjectedNet     float64 `json:"projected_net"`
	ProjectedBalance float64 `json:"projected_balance"`
}

// ForecastResult is the full forecast response payload.
type ForecastResult struct {
	Horizon     int             `json:"horizon"`
	Currency    string          `json:"currency"`
	Points      []ForecastPoint `json:"points"`
	Assumptions []string        `json:"assumptions"`
}

// frequencyMonthlyMultiplier returns approximate monthly occurrences for a frequency.
func frequencyMonthlyMultiplier(freq string) float64 {
	switch freq {
	case "weekly":
		return 52.0 / 12.0 // ~4.33
	case "biweekly":
		return 26.0 / 12.0 // ~2.17
	case "monthly":
		return 1.0
	case "quarterly":
		return 1.0 / 3.0
	case "yearly":
		return 1.0 / 12.0
	default:
		return 1.0
	}
}

// projectRecurring sums up the expected monthly recurring inflows and outflows.
// recurring rows without an expected_amount are skipped.
func projectRecurring(rows []RecurringRow) (inflow, outflow float64, merchants []string) {
	seen := make(map[string]bool)
	for _, r := range rows {
		if !r.ExpectedAmount.Valid || r.ExpectedAmount.Float64 <= 0 {
			continue
		}
		mult := frequencyMonthlyMultiplier(r.Frequency)
		monthly := r.ExpectedAmount.Float64 * mult
		// Heuristic: if merchant name hints at income (subscription, salary) treat as
		// inflow; for now use amount direction from category.  We lack a direction field
		// on recurring_transactions so we treat all recurring as outflows (expenses)
		// unless the caller overrides.  The assumption is surfaced in the response.
		outflow += monthly
		if !seen[r.MerchantNormalized] {
			merchants = append(merchants, r.MerchantNormalized)
			seen[r.MerchantNormalized] = true
		}
	}
	return 0, outflow, merchants
}

// ComputeForecast builds a horizon-month cash-flow projection.
// history must be ordered oldest → newest.
func ComputeForecast(
	history []MonthlyTotals,
	recurring []RecurringRow,
	horizon int,
	currency string,
) ForecastResult {
	if horizon < 1 {
		horizon = 3
	}
	if horizon > 24 {
		horizon = 24
	}

	// Compute historical averages.
	var totalIn, totalOut float64
	for _, h := range history {
		totalIn += h.In
		totalOut += h.Out
	}
	n := float64(len(history))
	avgIn, avgOut := 0.0, 0.0
	if n > 0 {
		avgIn = totalIn / n
		avgOut = totalOut / n
	}

	// Recurring contribution (monthly).
	_, recurOut, recurMerchants := projectRecurring(recurring)

	// Blend: use average as base, add recurring outflow on top of average if
	// average outflow is less than recurring (avoid double-counting if averages
	// already include these charges).
	blendedOut := avgOut
	if recurOut > avgOut {
		blendedOut = recurOut
	}
	blendedIn := avgIn

	// Build projection points.
	now := time.Now().UTC()
	startYear, startMonth, _ := now.Date()
	// project from next month
	projStart := time.Date(startYear, startMonth+1, 1, 0, 0, 0, 0, time.UTC)

	points := make([]ForecastPoint, 0, horizon)
	balance := 0.0 // running balance; starts at 0 (relative change)
	for i := 0; i < horizon; i++ {
		t := projStart.AddDate(0, i, 0)
		label := fmt.Sprintf("%d-%02d", t.Year(), t.Month())
		net := blendedIn - blendedOut
		balance += net
		points = append(points, ForecastPoint{
			Month:            label,
			ProjectedInflow:  roundTwo(blendedIn),
			ProjectedOutflow: roundTwo(blendedOut),
			ProjectedNet:     roundTwo(net),
			ProjectedBalance: roundTwo(balance),
		})
	}

	// Surface assumptions.
	assumptions := []string{}
	if len(history) == 0 {
		assumptions = append(assumptions, "No transaction history available; projection uses zero baseline")
	} else {
		assumptions = append(assumptions, fmt.Sprintf(
			"Historical averages computed from %d month(s) of data", len(history),
		))
	}
	if len(recurMerchants) > 0 {
		assumptions = append(assumptions, fmt.Sprintf(
			"%d recurring merchant(s) included: %s", len(recurMerchants), joinMerchants(recurMerchants, 5),
		))
	} else {
		assumptions = append(assumptions, "No active recurring transactions found")
	}
	assumptions = append(assumptions, "Recurring transactions treated as outflows (expense); override if merchant is income source")
	assumptions = append(assumptions, "Projection uses flat average; seasonality not modelled")

	return ForecastResult{
		Horizon:     horizon,
		Currency:    currency,
		Points:      points,
		Assumptions: assumptions,
	}
}

func joinMerchants(ms []string, max int) string {
	sort.Strings(ms)
	if len(ms) > max {
		ms = ms[:max]
		return fmt.Sprintf("%v and more", ms)
	}
	s := ""
	for i, m := range ms {
		if i > 0 {
			s += ", "
		}
		s += m
	}
	return s
}

// ─── Anomaly detection ────────────────────────────────────────────────────────

// AnomalyType enumerates the kinds of anomalies we detect.
type AnomalyType string

const (
	AnomalyDuplicate      AnomalyType = "duplicate"
	AnomalyUnusualSpend   AnomalyType = "unusual_spend"
	AnomalyMissingReceipt AnomalyType = "missing_receipt"
)

// Severity ranks anomalies.
type Severity string

const (
	SeverityHigh   Severity = "high"
	SeverityMedium Severity = "medium"
	SeverityLow    Severity = "low"
)

// Anomaly is one detected anomaly.
type Anomaly struct {
	ID            string      `json:"id"`
	Type          AnomalyType `json:"type"`
	Severity      Severity    `json:"severity"`
	Title         string      `json:"title"`
	Description   string      `json:"description"`
	Amount        *float64    `json:"amount,omitempty"`
	Currency      *string     `json:"currency,omitempty"`
	TransactionID *string     `json:"transaction_id,omitempty"`
	DetectedAt    string      `json:"detected_at"`
}

// highValueThreshold is the amount above which a transaction lacking a
// reconciled document is flagged as missing_receipt.
const highValueThreshold = 500.0

// duplicateDateWindowDays: two transactions are "near date" if within this window.
const duplicateDateWindowDays = 3

// DetectDuplicates finds pairs of transactions that share merchant_normalized +
// similar amount within a short date window.  Returns one anomaly per suspect pair.
func DetectDuplicates(txs []TxRow, detectedAt time.Time) []Anomaly {
	type key struct {
		merchant string
		amount   int64 // rounded to cents
	}
	// Group by key.
	type entry struct {
		id   uuid.UUID
		date time.Time
		amt  float64
		cur  string
	}
	groups := make(map[key][]entry)
	for _, t := range txs {
		if !t.MerchantNormalized.Valid || !t.Amount.Valid || !t.PostedDate.Valid {
			continue
		}
		k := key{
			merchant: t.MerchantNormalized.String,
			amount:   int64(math.Round(t.Amount.Float64 * 100)),
		}
		cur := ""
		if t.Currency.Valid {
			cur = t.Currency.String
		}
		groups[k] = append(groups[k], entry{
			id:   t.ID,
			date: t.PostedDate.Time,
			amt:  t.Amount.Float64,
			cur:  cur,
		})
	}

	var anomalies []Anomaly
	seen := make(map[string]bool)
	for _, entries := range groups {
		if len(entries) < 2 {
			continue
		}
		// Sort by date.
		sort.Slice(entries, func(i, j int) bool {
			return entries[i].date.Before(entries[j].date)
		})
		// Find pairs within date window.
		for i := 0; i < len(entries); i++ {
			for j := i + 1; j < len(entries); j++ {
				delta := entries[j].date.Sub(entries[i].date)
				if delta < 0 {
					delta = -delta
				}
				if delta > time.Duration(duplicateDateWindowDays)*24*time.Hour {
					break // sorted, remaining entries only get further away
				}
				pairKey := entries[i].id.String() + "+" + entries[j].id.String()
				if seen[pairKey] {
					continue
				}
				seen[pairKey] = true
				amt := entries[i].amt
				cur := entries[i].cur
				txID := entries[j].id.String() // flag the later one as suspect
				anomalies = append(anomalies, Anomaly{
					ID:            "dup-" + entries[i].id.String()[:8],
					Type:          AnomalyDuplicate,
					Severity:      SeverityHigh,
					Title:         "Possible duplicate transaction",
					Description:   fmt.Sprintf("Two transactions of %.2f from %q within %d days", amt, entries[i].id.String()[:8], duplicateDateWindowDays),
					Amount:        &amt,
					Currency:      &cur,
					TransactionID: &txID,
					DetectedAt:    detectedAt.Format(time.RFC3339),
				})
			}
		}
	}
	return anomalies
}

// DetectUnusualSpend uses a per-category z-score approach: a transaction whose
// amount is more than zThreshold standard deviations above the category mean is
// flagged.  history is map[catID] → []monthly_amounts.
func DetectUnusualSpend(txs []TxRow, history map[string][]float64, detectedAt time.Time) []Anomaly {
	const zThreshold = 2.5

	// Compute per-category mean and stddev.
	type stats struct {
		mean   float64
		stddev float64
	}
	catStats := make(map[string]stats)
	for catID, vals := range history {
		if len(vals) < 2 {
			continue // need at least 2 data points for stddev
		}
		mean := mean(vals)
		sd := stddev(vals, mean)
		if sd < 1.0 {
			sd = 1.0 // floor to avoid division by near-zero
		}
		catStats[catID] = stats{mean, sd}
	}

	var anomalies []Anomaly
	for _, t := range txs {
		if !t.CategoryID.Valid || !t.Amount.Valid || t.Direction != "debit" {
			continue
		}
		catID := t.CategoryID.UUID.String()
		cs, ok := catStats[catID]
		if !ok {
			continue
		}
		z := (t.Amount.Float64 - cs.mean) / cs.stddev
		if z < zThreshold {
			continue
		}
		sev := SeverityMedium
		if z > 4.0 {
			sev = SeverityHigh
		}
		amt := t.Amount.Float64
		cur := ""
		if t.Currency.Valid {
			cur = t.Currency.String
		}
		txID := t.ID.String()
		anomalies = append(anomalies, Anomaly{
			ID:            "usp-" + t.ID.String()[:8],
			Type:          AnomalyUnusualSpend,
			Severity:      sev,
			Title:         "Unusual spend in category",
			Description:   fmt.Sprintf("Transaction of %.2f is %.1f standard deviations above category average (%.2f)", amt, z, cs.mean),
			Amount:        &amt,
			Currency:      &cur,
			TransactionID: &txID,
			DetectedAt:    detectedAt.Format(time.RFC3339),
		})
	}
	return anomalies
}

// DetectMissingReceipts flags high-value debit transactions that lack a
// reconciled document match.
func DetectMissingReceipts(txs []TxRow, reconciledIDs map[uuid.UUID]struct{}, detectedAt time.Time) []Anomaly {
	var anomalies []Anomaly
	for _, t := range txs {
		if t.Direction != "debit" {
			continue
		}
		if !t.Amount.Valid || t.Amount.Float64 < highValueThreshold {
			continue
		}
		if _, ok := reconciledIDs[t.ID]; ok {
			continue
		}
		amt := t.Amount.Float64
		cur := ""
		if t.Currency.Valid {
			cur = t.Currency.String
		}
		txID := t.ID.String()
		anomalies = append(anomalies, Anomaly{
			ID:            "rcpt-" + t.ID.String()[:8],
			Type:          AnomalyMissingReceipt,
			Severity:      SeverityMedium,
			Title:         "High-value transaction without reconciled document",
			Description:   fmt.Sprintf("Transaction of %.2f has no confirmed or auto-matched document", amt),
			Amount:        &amt,
			Currency:      &cur,
			TransactionID: &txID,
			DetectedAt:    detectedAt.Format(time.RFC3339),
		})
	}
	return anomalies
}

// ─── Tax-readiness ────────────────────────────────────────────────────────────

// TaxComponent is one line in the tax-readiness component list.
type TaxComponent struct {
	Label  string `json:"label"`
	Status string `json:"status"` // "ok" | "warn" | "error"
	Detail string `json:"detail"`
}

// TaxReadinessResult is the full tax-readiness response payload.
type TaxReadinessResult struct {
	Score                float64        `json:"score"`
	VATPosition          *float64       `json:"vat_position,omitempty"`
	DocumentedExpensePct float64        `json:"documented_expense_pct"`
	UnreconciledCount    int            `json:"unreconciled_count"`
	Components           []TaxComponent `json:"components"`
}

// ComputeTaxReadiness derives a 0–100 score from three components:
//  1. VAT position (output − input): 40 pts — 40 if ≥ 0 (net payable tracked),
//     20 if data present but negative (refund position), 0 if no VAT data.
//  2. Document coverage for expenses: 40 pts × documented_pct.
//  3. Unreconciled penalty: max 20 pts, reduced by 1 per 10 unreconciled items.
func ComputeTaxReadiness(data *TaxReadinessData) TaxReadinessResult {
	var components []TaxComponent

	// Component 1: VAT tracking.
	vatScore := 0.0
	var vatPosition *float64
	hasVAT := data.VATOutput > 0 || data.VATInput > 0
	if hasVAT {
		pos := data.VATOutput - data.VATInput
		vatPosition = &pos
		if pos >= 0 {
			vatScore = 40.0
			components = append(components, TaxComponent{
				Label:  "VAT position",
				Status: "ok",
				Detail: fmt.Sprintf("Net VAT payable: %.2f (output %.2f, input %.2f)", pos, data.VATOutput, data.VATInput),
			})
		} else {
			vatScore = 20.0
			components = append(components, TaxComponent{
				Label:  "VAT position",
				Status: "warn",
				Detail: fmt.Sprintf("Net VAT refund position: %.2f — verify input tax claims", pos),
			})
		}
	} else {
		vatScore = 0.0
		components = append(components, TaxComponent{
			Label:  "VAT position",
			Status: "warn",
			Detail: "No VAT-tagged transactions found; assign tax rates to enable VAT tracking",
		})
	}

	// Component 2: document coverage.
	docPct := 0.0
	if data.TotalExpenses > 0 {
		docPct = float64(data.DocumentedExpenses) / float64(data.TotalExpenses) * 100
	}
	docScore := docPct / 100 * 40
	docStatus := "ok"
	if docPct < 50 {
		docStatus = "error"
	} else if docPct < 80 {
		docStatus = "warn"
	}
	components = append(components, TaxComponent{
		Label:  "Expense documentation",
		Status: docStatus,
		Detail: fmt.Sprintf("%.0f%% of expense transactions have a supporting document (%d of %d)", docPct, data.DocumentedExpenses, data.TotalExpenses),
	})

	// Component 3: reconciliation.
	reconScore := 20.0 - float64(data.UnreconciledCount)/10.0
	if reconScore < 0 {
		reconScore = 0
	}
	reconStatus := "ok"
	if data.UnreconciledCount > 50 {
		reconStatus = "error"
	} else if data.UnreconciledCount > 10 {
		reconStatus = "warn"
	}
	components = append(components, TaxComponent{
		Label:  "Reconciliation",
		Status: reconStatus,
		Detail: fmt.Sprintf("%d expense transaction(s) lack a reconciled bank statement line", data.UnreconciledCount),
	})

	score := vatScore + docScore + reconScore
	if score > 100 {
		score = 100
	}
	if score < 0 {
		score = 0
	}

	return TaxReadinessResult{
		Score:                roundTwo(score),
		VATPosition:          vatPosition,
		DocumentedExpensePct: roundTwo(docPct),
		UnreconciledCount:    data.UnreconciledCount,
		Components:           components,
	}
}

// ─── math helpers ─────────────────────────────────────────────────────────────

func mean(vals []float64) float64 {
	if len(vals) == 0 {
		return 0
	}
	var s float64
	for _, v := range vals {
		s += v
	}
	return s / float64(len(vals))
}

func stddev(vals []float64, m float64) float64 {
	if len(vals) < 2 {
		return 0
	}
	var s float64
	for _, v := range vals {
		d := v - m
		s += d * d
	}
	return math.Sqrt(s / float64(len(vals)))
}

func roundTwo(f float64) float64 {
	shifted := f * 100
	if shifted < 0 {
		shifted -= 0.5
	} else {
		shifted += 0.5
	}
	return float64(int64(shifted)) / 100
}

File: backend/src/modules/intelligence/compute.ts
/**
 * Intelligence compute — port of Go backend/internal/intelligence/compute.go.
 * Pure functions: no I/O, no side effects. Safe to unit-test directly.
 */
import type {
  MonthlyTotals,
  RecurringRow,
  ForecastPoint,
  ForecastResult,
  TxRow,
  Anomaly,
  TaxReadinessData,
  TaxReadinessResult,
  TaxComponent,
} from "./types";

// ─── Math helpers ─────────────────────────────────────────────────────────────

/** Port of Go roundTwo — round to 2 decimal places (same half-up logic). */
export function roundTwo(f: number): number {
  const shifted = f * 100;
  const rounded = shifted < 0 ? Math.ceil(shifted - 0.5) : Math.floor(shifted + 0.5);
  return rounded / 100;
}

function mean(vals: number[]): number {
  if (vals.length === 0) return 0;
  let s = 0;
  for (const v of vals) s += v;
  return s / vals.length;
}

function stddev(vals: number[], m: number): number {
  if (vals.length < 2) return 0;
  let s = 0;
  for (const v of vals) {
    const d = v - m;
    s += d * d;
  }
  return Math.sqrt(s / vals.length);
}

// ─── Forecast ─────────────────────────────────────────────────────────────────

/** Port of Go frequencyMonthlyMultiplier. */
export function frequencyMonthlyMultiplier(freq: string): number {
  switch (freq) {
    case "weekly":    return 52.0 / 12.0;  // ~4.33
    case "biweekly":  return 26.0 / 12.0;  // ~2.17
    case "monthly":   return 1.0;
    case "quarterly": return 1.0 / 3.0;
    case "yearly":    return 1.0 / 12.0;
    default:          return 1.0;
  }
}

/** Port of Go joinMerchants — sorts, caps at max, joins with ", ". */
function joinMerchants(ms: string[], max: number): string {
  const sorted = [...ms].sort();
  if (sorted.length > max) {
    return `${sorted.slice(0, max).join(", ")} and more`;
  }
  return sorted.join(", ");
}

/**
 * Port of Go projectRecurring — sums monthly recurring outflows.
 * Recurring rows without an expected_amount are skipped.
 * All recurring treated as outflows (Go assumption surfaced in response).
 */
export function projectRecurring(rows: RecurringRow[]): { outflow: number; merchants: string[] } {
  const seen = new Set<string>();
  const merchants: string[] = [];
  let outflow = 0;
  for (const r of rows) {
    if (r.expected_amount == null || r.expected_amount <= 0) continue;
    const mult = frequencyMonthlyMultiplier(r.frequency);
    const monthly = r.expected_amount * mult;
    outflow += monthly;
    if (!seen.has(r.merchant_normalized)) {
      seen.add(r.merchant_normalized);
      merchants.push(r.merchant_normalized);
    }
  }
  return { outflow, merchants };
}

/**
 * Port of Go ComputeForecast — builds horizon-month cash-flow projection.
 * history must be ordered oldest → newest.
 */
export function computeForecast(
  history: MonthlyTotals[],
  recurring: RecurringRow[],
  horizon: number,
  currency: string,
): ForecastResult {
  if (horizon < 1) horizon = 3;
  if (horizon > 24) horizon = 24;

  // Historical averages.
  let totalIn = 0;
  let totalOut = 0;
  for (const h of history) {
    totalIn += h.in;
    totalOut += h.out;
  }
  const n = history.length;
  let avgIn = 0;
  let avgOut = 0;
  if (n > 0) {
    avgIn = totalIn / n;
    avgOut = totalOut / n;
  }

  // Recurring contribution.
  const { outflow: recurOut, merchants: recurMerchants } = projectRecurring(recurring);

  // Blend: use average as base; if recurring outflow > avg outflow, use recurring.
  const blendedOut = recurOut > avgOut ? recurOut : avgOut;
  const blendedIn = avgIn;

  // Build projection points — project from next month.
  const now = new Date();
  const startYear = now.getUTCFullYear();
  const startMonth = now.getUTCMonth() + 1; // 1-based

  // next month
  let projYear = startYear;
  let projMonth = startMonth + 1;
  if (projMonth > 12) { projMonth = 1; projYear++; }

  const points: ForecastPoint[] = [];
  let balance = 0.0;

  for (let i = 0; i < horizon; i++) {
    // month label
    let y = projYear;
    let m = projMonth + i;
    while (m > 12) { m -= 12; y++; }
    const label = `${y}-${String(m).padStart(2, "0")}`;

    const net = blendedIn - blendedOut;
    balance += net;
    points.push({
      month: label,
      projected_inflow: roundTwo(blendedIn),
      projected_outflow: roundTwo(blendedOut),
      projected_net: roundTwo(net),
      projected_balance: roundTwo(balance),
    });
  }

  // Surface assumptions.
  const assumptions: string[] = [];
  if (history.length === 0) {
    assumptions.push("No transaction history available; projection uses zero baseline");
  } else {
    assumptions.push(`Historical averages computed from ${history.length} month(s) of data`);
  }
  if (recurMerchants.length > 0) {
    assumptions.push(`${recurMerchants.length} recurring merchant(s) included: ${joinMerchants(recurMerchants, 5)}`);
  } else {
    assumptions.push("No active recurring transactions found");
  }
  assumptions.push("Recurring transactions treated as outflows (expense); override if merchant is income source");
  assumptions.push("Projection uses flat average; seasonality not modelled");

  return { horizon, currency, points, assumptions };
}

// ─── Anomaly detection ────────────────────────────────────────────────────────

const HIGH_VALUE_THRESHOLD = 500.0;
const DUPLICATE_DATE_WINDOW_DAYS = 3;

/**
 * Port of Go DetectDuplicates — finds pairs sharing merchant + amount within date window.
 */
export function detectDuplicates(txs: TxRow[], detectedAt: Date): Anomaly[] {
  // Group by merchant + amount (rounded to cents).
  const groups = new Map<string, { id: string; date: Date; amt: number; cur: string }[]>();

  for (const t of txs) {
    if (t.merchant_normalized == null || t.amount == null || t.posted_date == null) continue;
    const amtCents = Math.round(t.amount * 100);
    const key = `${t.merchant_normalized}:${amtCents}`;
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key)!.push({
      id: t.id,
      date: new Date(t.posted_date),
      amt: t.amount,
      cur: t.currency ?? "",
    });
  }

  const anomalies: Anomaly[] = [];
  const seen = new Set<string>();
  const windowMs = DUPLICATE_DATE_WINDOW_DAYS * 24 * 60 * 60 * 1000;

  for (const entries of groups.values()) {
    if (entries.length < 2) continue;
    // Sort by date ascending.
    entries.sort((a, b) => a.date.getTime() - b.date.getTime());

    for (let i = 0; i < entries.length; i++) {
      for (let j = i + 1; j < entries.length; j++) {
        const delta = Math.abs(entries[j].date.getTime() - entries[i].date.getTime());
        if (delta > windowMs) break; // sorted, no point continuing

        const pairKey = `${entries[i].id}+${entries[j].id}`;
        if (seen.has(pairKey)) continue;
        seen.add(pairKey);

        const amt = entries[i].amt;
        const cur = entries[i].cur;
        const txId = entries[j].id; // flag the later one
        anomalies.push({
          id: `dup-${entries[i].id.slice(0, 8)}`,
          type: "duplicate",
          severity: "high",
          title: "Possible duplicate transaction",
          description: `Two transactions of ${amt.toFixed(2)} from "${entries[i].id.slice(0, 8)}" within ${DUPLICATE_DATE_WINDOW_DAYS} days`,
          amount: amt,
          currency: cur,
          transaction_id: txId,
          detected_at: detectedAt.toISOString(),
        });
      }
    }
  }
  return anomalies;
}

/**
 * Port of Go DetectUnusualSpend — z-score per category, threshold 2.5.
 * history: Map of categoryID → monthly amounts.
 */
export function detectUnusualSpend(
  txs: TxRow[],
  history: Map<string, number[]>,
  detectedAt: Date,
): Anomaly[] {
  const Z_THRESHOLD = 2.5;

  // Compute per-category stats.
  const catStats = new Map<string, { mean: number; stddev: number }>();
  for (const [catId, vals] of history.entries()) {
    if (vals.length < 2) continue; // need at least 2 points
    const m = mean(vals);
    let sd = stddev(vals, m);
    if (sd < 1.0) sd = 1.0; // floor
    catStats.set(catId, { mean: m, stddev: sd });
  }

  const anomalies: Anomaly[] = [];
  for (const t of txs) {
    if (t.category_id == null || t.amount == null || t.direction !== "debit") continue;
    const cs = catStats.get(t.category_id);
    if (!cs) continue;
    const z = (t.amount - cs.mean) / cs.stddev;
    if (z < Z_THRESHOLD) continue;

    const sev = z > 4.0 ? "high" : "medium";
    anomalies.push({
      id: `usp-${t.id.slice(0, 8)}`,
      type: "unusual_spend",
      severity: sev,
      title: "Unusual spend in category",
      description: `Transaction of ${t.amount.toFixed(2)} is ${z.toFixed(1)} standard deviations above category average (${cs.mean.toFixed(2)})`,
      amount: t.amount,
      currency: t.currency ?? undefined,
      transaction_id: t.id,
      detected_at: detectedAt.toISOString(),
    });
  }
  return anomalies;
}

/**
 * Port of Go DetectMissingReceipts — high-value debits without reconciled docs.
 */
export function detectMissingReceipts(
  txs: TxRow[],
  reconciledIds: Set<string>,
  detectedAt: Date,
): Anomaly[] {
  const anomalies: Anomaly[] = [];
  for (const t of txs) {
    if (t.direction !== "debit") continue;
    if (t.amount == null || t.amount < HIGH_VALUE_THRESHOLD) continue;
    if (reconciledIds.has(t.id)) continue;

    anomalies.push({
      id: `rcpt-${t.id.slice(0, 8)}`,
      type: "missing_receipt",
      severity: "medium",
      title: "High-value transaction without reconciled document",
      description: `Transaction of ${t.amount.toFixed(2)} has no confirmed or auto-matched document`,
      amount: t.amount,
      currency: t.currency ?? undefined,
      transaction_id: t.id,
      detected_at: detectedAt.toISOString(),
    });
  }
  return anomalies;
}

// ─── Tax readiness ────────────────────────────────────────────────────────────

/**
 * Port of Go ComputeTaxReadiness — 0–100 score from three components.
 *   1. VAT position: 40 pts (40 if net ≥ 0, 20 if net < 0, 0 if no VAT data)
 *   2. Doc coverage: 40 pts × documented_pct
 *   3. Reconciliation penalty: max 20 pts, reduced by 1 per 10 unreconciled
 */
export function computeTaxReadiness(data: TaxReadinessData): TaxReadinessResult {
  const components: TaxComponent[] = [];

  // Component 1: VAT tracking.
  let vatScore = 0.0;
  let vatPosition: number | undefined;
  const hasVAT = data.vat_output > 0 || data.vat_input > 0;
  if (hasVAT) {
    const pos = data.vat_output - data.vat_input;
    vatPosition = pos;
    if (pos >= 0) {
      vatScore = 40.0;
      components.push({
        label: "VAT position",
        status: "ok",
        detail: `Net VAT payable: ${pos.toFixed(2)} (output ${data.vat_output.toFixed(2)}, input ${data.vat_input.toFixed(2)})`,
      });
    } else {
      vatScore = 20.0;
      components.push({
        label: "VAT position",
        status: "warn",
        detail: `Net VAT refund position: ${pos.toFixed(2)} — verify input tax claims`,
      });
    }
  } else {
    vatScore = 0.0;
    components.push({
      label: "VAT position",
      status: "warn",
      detail: "No VAT-tagged transactions found; assign tax rates to enable VAT tracking",
    });
  }

  // Component 2: Document coverage.
  let docPct = 0.0;
  if (data.total_expenses > 0) {
    docPct = (data.documented_expenses / data.total_expenses) * 100;
  }
  const docScore = (docPct / 100) * 40;
  const docStatus: "ok" | "warn" | "error" = docPct < 50 ? "error" : docPct < 80 ? "warn" : "ok";
  components.push({
    label: "Expense documentation",
    status: docStatus,
    detail: `${Math.round(docPct)}% of expense transactions have a supporting document (${data.documented_expenses} of ${data.total_expenses})`,
  });

  // Component 3: Reconciliation.
  let reconScore = 20.0 - data.unreconciled_count / 10.0;
  if (reconScore < 0) reconScore = 0;
  const reconStatus: "ok" | "warn" | "error" =
    data.unreconciled_count > 50 ? "error" :
    data.unreconciled_count > 10 ? "warn" : "ok";
  components.push({
    label: "Reconciliation",
    status: reconStatus,
    detail: `${data.unreconciled_count} expense transaction(s) lack a reconciled bank statement line`,
  });

  let score = vatScore + docScore + reconScore;
  if (score > 100) score = 100;
  if (score < 0) score = 0;

  const result: TaxReadinessResult = {
    score: roundTwo(score),
    documented_expense_pct: roundTwo(docPct),
    unreconciled_count: data.unreconciled_count,
    components,
  };
  if (vatPosition !== undefined) result.vat_position = vatPosition;
  return result;
}
